finetune/adapter.py

```diff
@@ -222,13 +224,22 @@ def train(
 
 @torch.no_grad()
 def validate(
-    fabric: L.Fabric, model: GPT, val_data: List[Dict], tokenizer: Tokenizer, longest_seq_length: int
+    fabric: L.Fabric,
+    model: GPT,
+    val_data: List[Dict],
+    tokenizer: Tokenizer,
+    longest_seq_length: int,
+    padded_vocab_size: int,
 ) -> torch.Tensor:
     fabric.print("Validating ...")
     model.eval()
     losses = torch.zeros(eval_iters)
     for k in range(eval_iters):
         input_ids, targets = get_batch(fabric, val_data, longest_seq_length)
+        if (input_ids > padded_vocab_size).any():
```

Comment:
Two things here.

- You can do `model.config.padded_vocab_size` and avoid passing it across all functions
- This creates a branch on the value of input_ids, which will be expensive under compilation. This could slow down validation under that setting. Perhaps it would be better doing this only during the sanity check

This comment applies to the other scripts too

Reply:
@carmocca 
A question: something like:
```python
if (input_ids.detach() > padded_vocab_size).any():
  ...
```
will not work?

> When we perform operations on a tensor, PyTorch builds a computation graph that tracks the operations performed on the tensor. **The detach() function is used to remove a tensor from the computation graph, making it a standalone tensor that is no longer linked to the graph**.

Yes, it will create a new tensor but `input_ids` should be quite small anyway so, in theory, the impact will be negligible.

Reply:
> You can do model.config.padded_vocab_size and avoid passing it across all functions

Ah yes, this makes things easier! Thanks

> This creates a branch on the value of input_ids, which will be expensive under compilation. This could slow down validation under that setting. Perhaps it would be better doing this only during the sanity check

@Andrei-Aksionov 

I think the computation graph is not the issue, it should already be detached because the `validate` function runs in a `torch.no_grad()` context. 

When I understood the @carmocca correctly, it's just that for loops get expanded, no matter whether the computation is part of the networks computation graph, when a model is compiled? I honestly don't know the details of what gets compiled or not.

Reply:
It means that you cannot fuse operations because you need to stop and look at the precise values in the tensor to know which branch to take. This is not the end of the world, but since it only needs to be checked once, I suggest that it's done during the sanity checking call

Reply:
>  it should already be detached because the validate function runs in a torch.no_grad() context.

It looks like there is a difference between `detach` and `torch.no_grad`:
https://saturncloud.io/blog/difference-between-detach-and-with-torchnograd-in-pytorch/

> It means that you cannot fuse operations because you need to stop and look at the precise values in the tensor to know which branch to take. 

Oh, that branching! :)
I initially thought about Computational Graph branching, while you've meant [branch prediction](https://en.wikipedia.org/wiki/Branch_predictor).

> since it only needs to be checked once, I suggest that it's done during the sanity checking call

Agree with that. And it is a more appropriate place for it.